lib/clients/base.js

```diff
@@ -0,0 +1,47 @@
+var async = require('async');
+
+/**
+ * A base implementation for all the clients
+ */
+var BaseClient = function BaseClient(opts) {
+  this.opts = opts || {};
+  var self = this;
+  this.sendQConcurrency = opts.sendQueueConcurrency || 5;
+  this.sendQ = async.queue(function(data, cb) {
+    self.doSend(data, cb);
+  }, this.sendQConcurrency);
+};
+
+//convert the data json object to an array of strings
+BaseClient.prototype.buildMessages = function(/** data */) {
+  throw new Error('not implemented');
+};
+
+//implement how the messages should be sent
+BaseClient.prototype.transport = function(/** message, options, cb */) {
```

Comment:
The way I see it, this method should always be overwritten in the sub class. If it's not overwritten, I think it should be treated as an runtime exception. Also there isn't really a special error handler for the callback.